metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -0,0 +1,294 @@
+    tree_linearization->push_back(node);
+    uint32_t depth = 0;
+    if (tree.count(node)) {
+        for (const TaxId &child : tree.at(node)) {
+            dfs_statistics(child, tree, tree_linearization);
+            tree_linearization->push_back(node);
+            if (node_depth[child] > depth) {
+                depth = node_depth[child];
+            }
+        }
+    }
+    node_depth[node] = depth + 1;
+}
+
+void TaxonomyClsAnno::rmq_preprocessing(const std::vector<TaxId> &tree_linearization) {
+    uint32_t num_rmq_rows = log2(tree_linearization.size()) + 1;
+
+    rmq_data.resize(num_rmq_rows);
+    for (uint32_t i = 0; i < num_rmq_rows; ++i) {
+        rmq_data[i].resize(tree_linearization.size());
+    }
+
+    // Copy tree_linearization to rmq[0].
+    for (uint32_t i = 0; i < tree_linearization.size(); ++i) {
+        rmq_data[0][i] = tree_linearization[i];
+    }
+
+    // Delta represents the size of the RMQ's sliding window (always a power of 2).
+    uint32_t delta = 1;
+    for (uint32_t row = 1; row < num_rmq_rows; ++row) {
+        for (uint32_t i = 0; i + delta < tree_linearization.size(); ++i) {
+            // rmq_data[row][i] covers an interval of size delta=2^row and returns the node with the maximal depth among positions [i, i+2^row-1] in the linearization.
+            // According to 'this->dfs_statistics()': node_depth[leaf] = 1 and node_depth[root] = maximum distance to a leaf.
+            if (node_depth[rmq_data[row - 1][i]] >
+                node_depth[rmq_data[row - 1][i + delta]]) {
+                rmq_data[row][i] = rmq_data[row - 1][i];
+            } else {
+                rmq_data[row][i] = rmq_data[row - 1][i + delta];
+            }
+        }
+        delta *= 2;
+    }
+}
+
+std::vector<TaxId> TaxonomyClsAnno::get_lca_taxids_for_seq(const std::string_view &sequence, bool reversed) const {
+    cerr << "Assign class not implemented reversed = " << reversed << "\n";
+    throw std::runtime_error("get_lca_taxids_for_seq TaxonomyClsAnno not implemented. Received seq size"
+                             + to_string(sequence.size()));
+}
+
+std::vector<TaxId> TaxonomyClsImportDB::get_lca_taxids_for_seq(const std::string_view &sequence, bool reversed) const {
+    cerr << "Assign class not implemented reversed = " << reversed << "\n";
+    throw std::runtime_error("get_lca_taxids_for_seq TaxonomyClsImportDB not implemented. Received seq size"
+                             + to_string(sequence.size()));
+}
+
+TaxId TaxonomyClsAnno::find_lca(const std::vector<TaxId> &taxids) const {
+    throw std::runtime_error("find_lca TaxonomyClsAnno not implemented. Received taxids size"
+                             + to_string(taxids.size()));
+}
+
+TaxId TaxonomyClsImportDB::find_lca(const std::vector<TaxId> &taxids) const {
+    throw std::runtime_error("find_lca TaxonomyClsImportDB not implemented. Received taxids size"
+                             + to_string(taxids.size()));
+}
```

Comment:
these will be implemented in the next PR. The error messages are not very suggestive, but I had to avoid the "unused variable" cpp error